### PR TITLE
setup the initial version of google-drive logic to upload a file

### PR DESCRIPTION
**Pull request by 0marSalah**
i used ejs just to testing the function we ned discuss how i the function will get the file
make sure before `yarn start` you are in google-drive ditrectory

### CODE REVIEW

File: google-drive/dist/server.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    const { code } = req.query;
+    const { tokens } = await oauth2Client.getToken(code);
+    oauth2Client.setCredentials(tokens);
+    res.redirect('/upload');
+});
+app.get('/upload', async (req, res) => {
+    const uploader = new GoogleDriveUploader(oauth2Client);
+    try {
+        await uploader.uploadFile({
+            filePath: path.join(__dirname, 'credentials.json'),
```
Comment (gvelez17) [inline google-drive/dist/server.js]: great, but lets not hardcode this path in the real code, this is good for a test harness, not in the actual code

File: google-drive/src/server.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  res.redirect('/upload');
+});
+
+app.get('/upload', async (req, res) => {
+  const uploader = new GoogleDriveUploader(oauth2Client);
+
+  try {
+    await uploader.uploadFile({
+      filePath: path.join(__dirname, 'credentials.json'),
+      mimeType: 'application/json'
```
Comment (gvelez17) [inline google-drive/src/server.ts]: oh also lets not checkin /dist, can you .gitignore it, we don't want to check in build targets

File: google-drive/src/uploader.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+class GoogleDriveUploader {
+  constructor(private auth: any) {}
+
+  public async uploadFile({
+    filePath,
+    mimeType,
+  }: UploadOptions): Promise<void> {
+    const drive = google.drive({ version: 'v3', auth: this.auth });
+    const fileMetadata = {
+      name: path.basename(filePath),
```
Comment (gvelez17) [inline google-drive/src/uploader.ts]: lets please separate how we get the file from the upload to google drive

there should be a separate methods to accept the file or stream and separate ones to push to the google drive api.  as it is now it conflates reading from a local file with pushing to google drive, so it won't be useful for connecting to front end, lets separate those please

**Review by gvelez17 — CHANGES_REQUESTED** [review]
needs refactor to separate how we get the data from pushing it to google drive
needs to be able to accept data from front end not only file
do not specify file paths in the main code that should be in test harness
do not check in dist
thanks for work so far!

**Comment by gvelez17** [comment]
oh also remove DCO it is such a pain to have to refactor because of it, kill that thing

File: README.md [hunk 4]
```diff
@@ -7,6 +7,7 @@ npm i OR yarn
 
 ## usage
 ```bash
```
Comment (gvelez17) [inline README.md:9]: uh we want to change storage based on env variable not whole chunk of code in one place, look up some better patterns to encapsulate this please thanks

**Review by gvelez17 — CHANGES_REQUESTED** [review]
So all the google drive related stuff should be in one library file

The server should be general

we want to encapsulate the minimal differences we need for each storaage type, the top level should not know or care except for env var or some switch

look up design patterns for encapsulation, ask ai about encapsulating storages like talk to it about what we plan

File: src/credentials.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (gvelez17) [inline src/credentials.json]: lets move this out of source into a fixture directory?  its for testing right?

File: src/models/StorageContext.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  async retrieve(id: string) {
+    return this.strategy.retrieve(id);
+  }
+
+  async delete(id: string) {
+    return this.strategy.delete(id);
+  }
+}
+
+class StorageFactory {
```
Comment (gvelez17) [inline src/models/StorageContext.ts:28]: great , this is good!

File: src/credentials.json (hunk 5 above)
Comment (0marSalah) [reply src/credentials.json]: yes it is

**Pull request merged**